numpy/lib/function_base.py

```diff
@@ -3612,17 +3786,81 @@ def _quantile_is_valid(q):
     return True
 
 
-def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
-                           interpolation='linear', keepdims=False):
+def _validate_weights(a, q, axis, weights):
+    if weights is None:
+        wgt = None
+    else:
+        a = asanyarray(a)
+        wgt = np.asanyarray(weights)
+
+        if issubclass(a.dtype.type, (np.integer, np.bool_)):
+            result_dtype = np.result_type(a.dtype, wgt.dtype, 'f8')
+        else:
+            result_dtype = np.result_type(a.dtype, wgt.dtype)
+
+        broadcastable = False
+        if a.ndim == wgt.ndim and a.shape != wgt.shape:
+            broadcastable = all([a_dim == w_dim or w_dim == 1
+                                 for a_dim, w_dim
+                                 in zip(a.shape, wgt.shape)])
+
+        if a.shape != wgt.shape and not broadcastable:
+            if axis is None:
+                raise TypeError(
+                    "Axis must be specified when shapes of a and weights "
+                    "differ and not broadcastable.")
+            if wgt.ndim != 1:
+                raise TypeError(
+                    "1D weights expected when shapes of a and weights differ "
+                    " and not broadcastable.")
+            if wgt.shape[0] != a.shape[axis]:
+                raise ValueError(
+                    "Length of weights not compatible with specified axis.")
+            if not np.issubdtype(wgt.dtype, np.number):
+                raise ValueError("All weight entries must be numeric.")
```

Comment:
Seems unreasonable - what about `dtype=object`?

Reply:
Can you give me an example?  Maybe I could put that in the test as well.

Reply:
Using `decimal.Decimal` weights

Reply:
Hmm, how would I catch weights that are not numbers then?  I have a test like this:
```
with assert_raises_regex(ValueError,
                                            'All weight entries must be numeric'):
    np.quantile(ar, q=0.5, axis=0, weights=[1, 'bad'])
```